File: frontend/src/api/speech.ts
import { apiRequest, resolveApiBase } from "./client"

// 一个预设服务商：只提供默认值，真正跑合成时仍以用户填写的字段为准。
export interface SpeechProviderSpec {
  id: string
  display_name: string
  base_url?: string
  model?: string
  voice?: string
  auth_style?: string
  local?: boolean
  voice_hint?: string
  endpoint_hint?: string
}

// 全局语音合成状态。密钥值永不回传，只回传"是否已配置"。
export interface SpeechStatus {
  provider: string
  base_url?: string
  model?: string
  voice?: string
  format?: string
  key_configured: boolean
  configured: boolean
  providers: SpeechProviderSpec[]
}

// 一个语音角色：可覆盖全局的接口地址／模型／发音人。
export interface VoiceRole {
  id: string
  name: string
  bio?: string
  has_avatar: boolean
  provider?: string
  base_url?: string
  model?: string
  voice?: string
  sort_order: number
  created_at: string
  updated_at: string
}

export interface SpeechSettingsResponse {
  speech: SpeechStatus
  roles: VoiceRole[]
  active_role_id: string
}

// 每个字段都是可选的：省略表示不改，传空字符串表示清空。
export interface SpeechConfigInput {
  provider?: string
  base_url?: string
  api_key?: string
  model?: string
  voice?: string
  format?: string
}

export interface VoiceRoleInput {
  name: string
  bio: string
  provider: string
  base_url: string
  model: string
  voice: string
  sort_order: number
}

// 更新走指针语义：只把用户真正改过的字段放进来。
export interface VoiceRolePatch {
  name?: string
  bio?: string
  provider?: string
  base_url?: string
  model?: string
  voice?: string
  sort_order?: number
}

export interface VoiceRoleListResponse {
  items: VoiceRole[]
  count: number
  active_role_id: string
}

export interface VoiceRoleAvatarResult {
  id: string
  has_avatar: boolean
  size_bytes: number
}

export function getSpeechSettings(): Promise<SpeechSettingsResponse> {
  return apiRequest<SpeechSettingsResponse>("/speech")
}

// 保存全局接口配置。API Key 写入本地 env 文件，响应里永远不会回显它的值。
export function saveSpeechConfig(input: SpeechConfigInput): Promise<{ speech: SpeechStatus }> {
  return apiRequest<{ speech: SpeechStatus }>("/speech/config", {
    method: "PATCH",
    body: JSON.stringify(input),
  })
}

export function listVoiceRoles(): Promise<VoiceRoleListResponse> {
  return apiRequest<VoiceRoleListResponse>("/speech/roles")
}

export function createVoiceRole(input: VoiceRoleInput): Promise<VoiceRole> {
  return apiRequest<VoiceRole>("/speech/roles", {
    method: "POST",
    body: JSON.stringify(input),
  })
}

export function updateVoiceRole(id: string, patch: VoiceRolePatch): Promise<VoiceRole> {
  return apiRequest<VoiceRole>(`/speech/roles/${encodeURIComponent(id)}`, {
    method: "PATCH",
    body: JSON.stringify(patch),
  })
}

export function deleteVoiceRole(id: string): Promise<void> {
  return apiRequest<void>(`/speech/roles/${encodeURIComponent(id)}`, { method: "DELETE" })
}

// 传空字符串表示取消选择，回到全局默认发音。
export function setActiveVoiceRole(roleID: string): Promise<{ active_role_id: string }> {
  return apiRequest<{ active_role_id: string }>("/speech/roles/active", {
    method: "PATCH",
    body: JSON.stringify({ role_id: roleID }),
  })
}

// apiRequest 遇到 FormData 会自动跳过 Content-Type，交给浏览器补 boundary。
export function uploadVoiceRoleAvatar(id: string, file: File): Promise<VoiceRoleAvatarResult> {
  const body = new FormData()
  body.append("file", file)
  return apiRequest<VoiceRoleAvatarResult>(`/speech/roles/${encodeURIComponent(id)}/avatar`, {
    method: "POST",
    body,
  })
}

// 头像要的是一个能直接喂给 <img src> 的绝对地址，所以走 resolveApiBase 而不是
// apiRequest。version 是缓存击穿参数：重传后同名文件会被浏览器认成旧图。
export function voiceRoleAvatarURL(id: string, version?: string | number): string {
  const url = `${resolveApiBase()}/speech/roles/${encodeURIComponent(id)}/avatar`
  return version === undefined ? url : `${url}?v=${encodeURIComponent(String(version))}`
}

// 生成路由要求带上这个头，缺了就是一声不响的 403 而不是一段音频。
const generationHeader = "X-Study-OS-Request"

// 试听拿的是音频字节而不是 JSON，所以绕开 apiRequest 直接 fetch。调用方负责
// 播放和 revokeObjectURL——这里只管把 blob 交出去。
export async function synthesizeVoiceRolePreview(id: string, text: string): Promise<Blob> {
  // 不写 format，让后端用设置里选的容器——试听要听的就是设置生效后的样子。
  const query = new URLSearchParams({ term: text, role: id })
  const response = await fetch(`${resolveApiBase()}/audio?${query.toString()}`, {
    method: "POST",
    headers: { [generationHeader]: "1" },
  })
  if (!response.ok) {
    // 后端在失败时回的是 {"error": "..."}，能读到就用它，读不到再退回状态码。
    let message = `试听失败（HTTP ${response.status}）`
    try {
      const payload = (await response.json()) as { error?: string }
      if (payload?.error) message = payload.error
    } catch {
      // 响应体不是 JSON，保留上面的兜底文案。
    }
    throw new Error(message)
  }
  return response.blob()
}

File: backend/httpapi/speech.go
package httpapi

import (
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"study-os/backend/app"
	"study-os/backend/config"
	"study-os/backend/models"
)

// maxVoiceAvatarBytes bounds a role avatar. Avatars render at roughly 40px, so
// anything larger is a user mistake rather than a requirement.
const maxVoiceAvatarBytes = 2 << 20

var allowedAvatarExtensions = map[string]string{
	".png":  "image/png",
	".jpg":  "image/jpeg",
	".jpeg": "image/jpeg",
	".gif":  "image/gif",
	".webp": "image/webp",
}

// handleSpeechSettings returns the 语音合成 endpoint configuration together with
// the saved voice roles, so the settings panel renders from a single request.
func handleSpeechSettings(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	roles, err := application.Store.ListVoiceRoles(request.Context())
	if err != nil {
		writeStoreError(response, err)
		return
	}
	activeID, err := application.Store.ActiveVoiceRoleID(request.Context())
	if err != nil {
		writeStoreError(response, err)
		return
	}
	writeJSON(response, http.StatusOK, map[string]any{
		"speech":         application.Config.SpeechStatus(),
		"roles":          roles,
		"active_role_id": activeID,
	})
}

// handleSpeechConfig persists the global endpoint to the local env file. The API
// key is written but never echoed back, matching how chat vendor keys behave.
func handleSpeechConfig(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	var input struct {
		Provider *string `json:"provider"`
		BaseURL  *string `json:"base_url"`
		APIKey   *string `json:"api_key"`
		Model    *string `json:"model"`
		Voice    *string `json:"voice"`
		Format   *string `json:"format"`
	}
	request.Body = http.MaxBytesReader(response, request.Body, maxImportJSONBytes)
	if err := decodeRequest(request, &input); err != nil {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": err.Error()})
		return
	}

	updates := make(map[string]string)
	if input.Provider != nil {
		provider := strings.TrimSpace(*input.Provider)
		if provider != "" {
			if _, ok := config.LookupSpeechProvider(provider); !ok {
				writeJSON(response, http.StatusBadRequest, map[string]string{"error": "unsupported speech provider"})
				return
			}
		}
		updates["SPEECH_PROVIDER"] = provider
	}
	if input.BaseURL != nil {
		updates["SPEECH_BASE_URL"] = strings.TrimSpace(*input.BaseURL)
	}
	if input.APIKey != nil {
		updates["SPEECH_API_KEY"] = strings.TrimSpace(*input.APIKey)
	}
	if input.Model != nil {
		updates["SPEECH_MODEL"] = strings.TrimSpace(*input.Model)
	}
	if input.Voice != nil {
		updates["SPEECH_VOICE"] = strings.TrimSpace(*input.Voice)
	}
	if input.Format != nil {
		format := strings.ToLower(strings.TrimSpace(*input.Format))
		if format != "" && !isSupportedSpeechFormat(format) {
			writeJSON(response, http.StatusBadRequest, map[string]string{"error": "unsupported audio format"})
			return
		}
		updates["SPEECH_FORMAT"] = format
	}
	if len(updates) == 0 {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": "no config fields provided"})
		return
	}
	if err := config.UpdateEnvFile(application.Config.EnvFilePath, updates); err != nil {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": err.Error()})
		return
	}

	// Mirror the write into the live config so synthesis picks it up without a
	// restart, then rebuild the generator around the new endpoint.
	settings := application.Config.SpeechSettings
	if value, ok := updates["SPEECH_PROVIDER"]; ok {
		settings.Provider = value
	}
	if value, ok := updates["SPEECH_BASE_URL"]; ok {
		settings.BaseURL = value
	}
	if value, ok := updates["SPEECH_API_KEY"]; ok {
		settings.APIKey = value
	}
	if value, ok := updates["SPEECH_MODEL"]; ok {
		settings.Model = value
	}
	if value, ok := updates["SPEECH_VOICE"]; ok {
		settings.Voice = value
	}
	if value, ok := updates["SPEECH_FORMAT"]; ok {
		settings.Format = value
	}
	application.Config.SpeechSettings = settings
	application.RebuildAudioGenerator()

	writeJSON(response, http.StatusOK, map[string]any{"speech": application.Config.SpeechStatus()})
}

func handleVoiceRoleList(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	roles, err := application.Store.ListVoiceRoles(request.Context())
	if err != nil {
		writeStoreError(response, err)
		return
	}
	activeID, _ := application.Store.ActiveVoiceRoleID(request.Context())
	writeJSON(response, http.StatusOK, map[string]any{
		"items": roles, "count": len(roles), "active_role_id": activeID,
	})
}

func handleVoiceRoleCreate(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	var input struct {
		Name      string `json:"name"`
		Bio       string `json:"bio"`
		Provider  string `json:"provider"`
		BaseURL   string `json:"base_url"`
		Model     string `json:"model"`
		Voice     string `json:"voice"`
		SortOrder int    `json:"sort_order"`
	}
	request.Body = http.MaxBytesReader(response, request.Body, maxImportJSONBytes)
	if err := decodeRequest(request, &input); err != nil {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": err.Error()})
		return
	}
	name := strings.TrimSpace(input.Name)
	if name == "" {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": "角色名称不能为空"})
		return
	}
	now := time.Now().UTC()
	role := models.VoiceRole{
		ID:        newRequestID("voice"),
		Name:      name,
		Bio:       strings.TrimSpace(input.Bio),
		Provider:  strings.TrimSpace(input.Provider),
		BaseURL:   strings.TrimSpace(input.BaseURL),
		Model:     strings.TrimSpace(input.Model),
		Voice:     strings.TrimSpace(input.Voice),
		SortOrder: input.SortOrder,
		CreatedAt: now,
		UpdatedAt: now,
	}
	if err := application.Store.CreateVoiceRole(request.Context(), role); err != nil {
		writeStoreError(response, err)
		return
	}
	writeJSON(response, http.StatusCreated, role)
}

func handleVoiceRoleUpdate(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	roleID := strings.TrimSpace(chi.URLParam(request, "roleID"))
	existing, err := application.Store.GetVoiceRole(request.Context(), roleID)
	if err != nil {
		writeStoreError(response, err)
		return
	}
	var input struct {
		Name      *string `json:"name"`
		Bio       *string `json:"bio"`
		Provider  *string `json:"provider"`
		BaseURL   *string `json:"base_url"`
		Model     *string `json:"model"`
		Voice     *string `json:"voice"`
		SortOrder *int    `json:"sort_order"`
	}
	request.Body = http.MaxBytesReader(response, request.Body, maxImportJSONBytes)
	if err := decodeRequest(request, &input); err != nil {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": err.Error()})
		return
	}
	if input.Name != nil {
		name := strings.TrimSpace(*input.Name)
		if name == "" {
			writeJSON(response, http.StatusBadRequest, map[string]string{"error": "角色名称不能为空"})
			return
		}
		existing.Name = name
	}
	if input.Bio != nil {
		existing.Bio = strings.TrimSpace(*input.Bio)
	}
	if input.Provider != nil {
		existing.Provider = strings.TrimSpace(*input.Provider)
	}
	if input.BaseURL != nil {
		existing.BaseURL = strings.TrimSpace(*input.BaseURL)
	}
	if input.Model != nil {
		existing.Model = strings.TrimSpace(*input.Model)
	}
	if input.Voice != nil {
		existing.Voice = strings.TrimSpace(*input.Voice)
	}
	if input.SortOrder != nil {
		existing.SortOrder = *input.SortOrder
	}
	if err := application.Store.UpdateVoiceRole(request.Context(), existing); err != nil {
		writeStoreError(response, err)
		return
	}
	updated, err := application.Store.GetVoiceRole(request.Context(), roleID)
	if err != nil {
		writeStoreError(response, err)
		return
	}
	writeJSON(response, http.StatusOK, updated)
}

func handleVoiceRoleDelete(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	roleID := strings.TrimSpace(chi.URLParam(request, "roleID"))
	existing, err := application.Store.GetVoiceRole(request.Context(), roleID)
	if err != nil {
		writeStoreError(response, err)
		return
	}
	if err := application.Store.DeleteVoiceRole(request.Context(), roleID); err != nil {
		writeStoreError(response, err)
		return
	}
	// The row is gone either way; a leftover avatar is only wasted disk, so a
	// failure here must not turn a successful delete into an error.
	if existing.AvatarPath != "" {
		_ = os.Remove(existing.AvatarPath)
	}
	response.WriteHeader(http.StatusNoContent)
}

func handleVoiceRoleActivate(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	var input struct {
		RoleID string `json:"role_id"`
	}
	request.Body = http.MaxBytesReader(response, request.Body, maxImportJSONBytes)
	if err := decodeRequest(request, &input); err != nil {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": err.Error()})
		return
	}
	if err := application.Store.SetActiveVoiceRole(request.Context(), strings.TrimSpace(input.RoleID)); err != nil {
		writeStoreError(response, err)
		return
	}
	writeJSON(response, http.StatusOK, map[string]any{"active_role_id": strings.TrimSpace(input.RoleID)})
}

func handleVoiceRoleAvatarUpload(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	roleID := strings.TrimSpace(chi.URLParam(request, "roleID"))
	existing, err := application.Store.GetVoiceRole(request.Context(), roleID)
	if err != nil {
		writeStoreError(response, err)
		return
	}
	request.Body = http.MaxBytesReader(response, request.Body, maxVoiceAvatarBytes+(1<<20))
	if err := request.ParseMultipartForm(maxVoiceAvatarBytes + (1 << 20)); err != nil {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": "无效的文件上传"})
		return
	}
	file, header, err := request.FormFile("file")
	if err != nil {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": "multipart file field is required"})
		return
	}
	defer file.Close()
	extension := strings.ToLower(filepath.Ext(header.Filename))
	if _, ok := allowedAvatarExtensions[extension]; !ok {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": "头像仅支持 PNG / JPG / GIF / WebP"})
		return
	}
	if header.Size > maxVoiceAvatarBytes {
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": "头像不能超过 2MB"})
		return
	}
	directory := filepath.Join(application.Config.DataDir, "uploads", "voice-roles")
	if err := os.MkdirAll(directory, 0o700); err != nil {
		writeJSON(response, http.StatusInternalServerError, map[string]string{"error": "创建头像目录失败"})
		return
	}
	destination := filepath.Join(directory, roleID+extension)
	out, err := os.Create(destination)
	if err != nil {
		writeJSON(response, http.StatusInternalServerError, map[string]string{"error": "保存头像失败"})
		return
	}
	written, copyErr := io.Copy(out, io.LimitReader(file, maxVoiceAvatarBytes+1))
	_ = out.Close()
	if copyErr != nil || written > maxVoiceAvatarBytes {
		_ = os.Remove(destination)
		writeJSON(response, http.StatusBadRequest, map[string]string{"error": "头像不能超过 2MB"})
		return
	}
	if err := application.Store.SetVoiceRoleAvatar(request.Context(), roleID, destination); err != nil {
		_ = os.Remove(destination)
		writeStoreError(response, err)
		return
	}
	// Replacing a PNG with a JPG leaves the old file behind under a different
	// extension, which would keep shadowing disk space forever.
	if existing.AvatarPath != "" && existing.AvatarPath != destination {
		_ = os.Remove(existing.AvatarPath)
	}
	writeJSON(response, http.StatusOK, map[string]any{"id": roleID, "has_avatar": true, "size_bytes": written})
}

func handleVoiceRoleAvatarGet(response http.ResponseWriter, request *http.Request, application *app.App) {
	if application == nil || application.Store == nil {
		writeJSON(response, http.StatusServiceUnavailable, map[string]string{"error": "application unavailable"})
		return
	}
	role, err := application.Store.GetVoiceRole(request.Context(), strings.TrimSpace(chi.URLParam(request, "roleID")))
	if err != nil {
		writeStoreError(response, err)
		return
	}
	if strings.TrimSpace(role.AvatarPath) == "" {
		writeJSON(response, http.StatusNotFound, map[string]string{"error": "该角色还没有头像"})
		return
	}
	if contentType, ok := allowedAvatarExtensions[strings.ToLower(filepath.Ext(role.AvatarPath))]; ok {
		response.Header().Set("Content-Type", contentType)
	}
	// The avatar changes in place when re-uploaded, so let the browser revalidate
	// rather than serve a stale face from cache.
	response.Header().Set("Cache-Control", "private, no-cache")
	http.ServeFile(response, request, role.AvatarPath)
}

func isSupportedSpeechFormat(format string) bool {
	switch format {
	case "wav", "mp3", "ogg", "m4a", "aac", "flac":
		return true
	default:
		return false
	}
}
